pkg/api/helpers.go

```diff
@@ -496,37 +496,81 @@ func GetTaintsFromNodeAnnotations(annotations map[string]string) ([]Taint, error
 }
 
 // TolerationToleratesTaint checks if the toleration tolerates the taint.
-func TolerationToleratesTaint(toleration *Toleration, taint *Taint) bool {
-	if len(toleration.Effect) != 0 && toleration.Effect != taint.Effect {
+func (t *Toleration) ToleratesTaint(taint *Taint) bool {
+	if len(t.Effect) != 0 && t.Effect != taint.Effect {
 		return false
 	}
 
-	if toleration.Key != taint.Key {
+	if t.Key != taint.Key {
 		return false
 	}
-	// TODO: Use proper defaulting when Toleration becomes a field of PodSpec
-	if (len(toleration.Operator) == 0 || toleration.Operator == TolerationOpEqual) && toleration.Value == taint.Value {
-		return true
+
+	// check forgivenessSeconds time out
+	if t.ForgivenessSeconds != nil {
```

Comment:
I assume that other nil "pointer to time" is interpreted as "not set." I think the forgivenessSeconds use case is different because if you didn't want to tolerate the taint, you wouldn't create a toleration in the first place. So there's no reason you would have a toleration with a nil forgivenessSeconds. So we can use it to represent infinite forgiveness. (Also, Kevin and I couldn't come up with any other reasonable, compact, and non-hacky (e.g. MaxInt or -1 or whatever) way to represent infinite forgiveness.)